### PR TITLE
Support for CST 820 Touch screen controller

### PR DESCRIPTION
**Pull request by dirkjankrijnders**
# What does this implement/fix?

Added support for the CST860 Touch screen controller as found on 2432S024C boards. Implementation is based on other implementations as a datasheet for this controller is not readily available
## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
touchscreen:
  platform: cst860
  id: my_touchscreen
  i2c_id: bus_a
  transform:
    mirror_y: false
    swap_xy: false
  on_touch:
      - lambda: |-
            ESP_LOGI("cal", "x=%d, y=%d, x_raw=%d, y_raw=%0d",
                touches[0].x,
                touches[0].y,
                touches[0].x_raw,
                touches[0].y_raw
                );
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by nielsnl68** [comment]
Hello @dirkjankrijnders, how are you doing? Recently i created a PR around the touchscreen component's, trying to make it more universal. (https://github.com/esphome/esphome/pull/4596) could you check it out and make the needed changes to your component so we can merge both at the same time?

**Comment by dirkjankrijnders** [comment]
Hi @nielsnl68, I've been looking through your PR and looks like a very useful change. But I'm a bit lost on how to change my PR. I guess forking yours and adapting my changes to that would work. I also have no issues waiting a bit for your PR to be merged and then following up on this one.

File: esphome/components/cst860/binary_sensor.py
Comment (nielsnl68) [inline esphome/components/cst860/binary_sensor.py:1]: You can remove this file.

File: esphome/components/cst860/cst860.cpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+#include <algorithm>
+#include <cinttypes>
+
+namespace esphome {
+namespace cst860 {
+
+static const char *const TAG = "cst860";
+
+void CST860TouchscreenStore::gpio_intr(CST860TouchscreenStore *store) { store->touch = true; }
```
Comment (nielsnl68) [inline esphome/components/cst860/cst860.cpp]:
```suggestion
```

File: esphome/components/cst860/cst860.cpp [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+void CST860Component::loop() {
+  if ((this->irq_pin_ != nullptr) && (this->store_.touch || this->touched)) {
+    this->store_.touch = false;
+    check_touch_();
+  }
+}
+
+void CST860Component::update() { check_touch_(); }
+
```
Comment (nielsnl68) [inline esphome/components/cst860/cst860.cpp]:
```suggestion
```

File: esphome/components/cst860/cst860.cpp [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+void CST860Component::check_touch_() {
```
Comment (nielsnl68) [inline esphome/components/cst860/cst860.cpp]:
```suggestion
void CST860Component::update_touches() {
```

File: esphome/components/cst860/cst860.cpp [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  if (err) {
+    ESP_LOGD("TAG", "Reg 0x2 Error: %i", err);
+  }
+  if (fingerindex == 0) {
+    if (this->touched) {
+      this->touched = false;
+      for (auto *listener : this->touch_listeners_) {
+        listener->release();
+      };
+    };
```
Comment (nielsnl68) [inline esphome/components/cst860/cst860.cpp]:
```suggestion
```

File: esphome/components/cst860/cst860.cpp [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+  } else {
+    touchpoint.y = this->y_raw_max_ - y;
+  };
+  if (!this->touched || (now - this->last_pos_ms_) >= this->report_millis_) {
+    this->defer([this, touchpoint]() { this->send_touch_(touchpoint); });
+    this->x = touchpoint.x;
+    this->y = touchpoint.y;
+    this->touched = true;
+    this->last_pos_ms_ = now;
+  };
```
Comment (nielsnl68) [inline esphome/components/cst860/cst860.cpp]:
```suggestion
    y = (((data[0] & 0x0f) << 8) | data[1]);
    x = (((data[2] & 0x0f) << 8) | data[3]);
    set_raw_touch_posistion_(0, x, y);
```

File: esphome/components/cst860/cst860.h [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+  void loop() override;
+
+  /** Read and process the values from the hardware.
+   *
+   * Read the raw x, y and touch pressure values from the chip, detect the touch,
+   * and if touched, transform to the user x and y coordinates. If the state has
+   * changed or if the value should be reported again due to the
+   * report interval, run the action and inform the virtual buttons.
+   */
+  void update() override;
```
Comment (nielsnl68) [inline esphome/components/cst860/cst860.h]:
```suggestion
```

File: esphome/components/cst860/cst860.h [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+  /// True if the component currently detects the touch
+  bool touched{false};
+
+  /**@{*/
+  /** Raw sensor values of the coordinates and the pressure.
+   *
+   * The values are set each time the update() method is called.
+   */
+  int16_t x_raw{0}, y_raw{0}, z_raw{0};
```
Comment (nielsnl68) [inline esphome/components/cst860/cst860.h]:
```suggestion
```

File: esphome/components/cst860/cst860.h [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+  /**@}*/
+
+ protected:
+  void check_touch_();
```
Comment (nielsnl68) [inline esphome/components/cst860/cst860.h]:
```suggestion
  void update_touches() override;
```

File: esphome/components/cst860/cst860.h [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+  int16_t x_raw_max_, y_raw_max_;
+
+  bool swap_x_y_, invert_x_, invert_y_;
+
+  uint32_t report_millis_;
+  uint32_t last_pos_ms_{0};
+
+  InternalGPIOPin *irq_pin_{nullptr};
+  CST860TouchscreenStore store_;
```
Comment (nielsnl68) [inline esphome/components/cst860/cst860.h]:
```suggestion
```

File: esphome/components/cst860/cst860.h [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+  static void gpio_intr(CST860TouchscreenStore *store);
+};
+
+class CST860Component : public Touchscreen, public PollingComponent, public i2c::I2CDevice {
+ public:
+  /// Set the logical touch screen dimensions.
+  void set_dimensions(int16_t x, int16_t y) {
+    this->display_width_ = x;
+    this->display_height_ = y;
+  }
```
Comment (nielsnl68) [inline esphome/components/cst860/cst860.h]:
```suggestion
```

File: esphome/components/cst860/cst860.h [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+  /// Set the coordinates for the touch screen edges.
+  void set_calibration(int16_t x_max, int16_t y_max, bool invert_x, bool invert_y);
+  /// If true the x and y axes will be swapped
+  void set_swap_x_y(bool val) { this->swap_x_y_ = val; }
+
+  /// Set the interval to report the touch point perodically.
+  void set_report_interval(uint32_t interval) { this->report_millis_ = interval; }
+  uint32_t get_report_interval() { return this->report_millis_; }
```
Comment (nielsnl68) [inline esphome/components/cst860/cst860.h]:
```suggestion
```

File: esphome/components/cst860/cst860.h [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+namespace esphome {
+namespace cst860 {
+
+using namespace touchscreen;
+
+struct CST860TouchscreenStore {
+  volatile bool touch;
+  static void gpio_intr(CST860TouchscreenStore *store);
+};
+
```
Comment (nielsnl68) [inline esphome/components/cst860/cst860.h]:
```suggestion
```

File: esphome/components/cst860/touchscreen.py [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+CONFIG_SCHEMA = touchscreen.TOUCHSCREEN_SCHEMA.extend(
+    cv.Schema(
+        {
+            cv.GenerateID(): cv.declare_id(CST860Component),
+            cv.Optional(CONF_INTERRUPT_PIN): cv.All(
+                pins.internal_gpio_input_pin_schema
+            ),
+            cv.Optional(CONF_X_MAX, default=4095): cv.int_range(min=0, max=4095),
+            cv.Optional(CONF_Y_MAX, default=4095): cv.int_range(min=0, max=4095),
+            cv.Optional(CONF_REPORT_INTERVAL, default="never"): report_interval,
```
Comment (nielsnl68) [inline esphome/components/cst860/touchscreen.py]:
```suggestion
```

File: esphome/components/cst860/touchscreen.py [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+).add_extra(validate_cst860)
+
+
+async def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    await cg.register_component(var, config)
+    await i2c.register_i2c_device(var, config)
+    await touchscreen.register_touchscreen(var, config)
+
+    cg.add(var.set_report_interval(config[CONF_REPORT_INTERVAL]))
```
Comment (nielsnl68) [inline esphome/components/cst860/touchscreen.py]:
```suggestion
```

**Review by nielsnl68 — COMMENTED** [review]
The following changes are based on the new touchscreen component. As you can see a lot of work has been taken out of your hands already by the Touchscreen class.

At the moment you do not need to update any of the changes. Until my PR https://github.com/esphome/esphome/pull/4596 is merged.

**Comment by nielsnl68** [comment]
Hallo @dirkjankrijnders do you have any code or documentation that you used to based this code on? I love to compare and help make this PR better,

**Comment by dirkjankrijnders** [comment]
I based this mostly on this [code](https://github.com/rzeldent/esp32-smartdisplay/blob/main/src/touch_cst820.cpp)

**Comment by dirkjankrijnders** [comment]
Another implementation is [here](https://github.com/Xinyuan-LilyGO/T-RGB/blob/main/lib/TouchLib/src/ModulesCSTSelf.tpp)

**Comment by dirkjankrijnders** [comment]
I spend some time digging further and it turns out the device actually is a CST820. The legend on the chip was badly readable... That led to a product page: http://en.hynitron.com/product/wearable/126. But the datasheet link there is dead, also on the Chinese version. However I found the datasheet for the CST816S:  https://www.buydisplay.com/download/ic/DS-CST816S_DS_V1.3.pdf. This seems pretty similar if I compare to the other implementations. I'll update this PR when I have found some time to process the new information and played with the new touchscreen implementation.

**Comment by nielsnl68** [comment]
> I based this mostly on this [code](https://github.com/rzeldent/esp32-smartdisplay/blob/main/src/touch_cst820.cpp)

Hi @dirkjankrijnders As of this month the touchscreen main class has got the update that i talked about earlier, please check out the dev or beta branch of esphome. The only thing you now need to do is setup the needed pins and override the `update_touches_()` method. and call `set_raw_touch_position_(id, raw_x, raw_y [, pressure]);` for every detected touch on the screen.

File: esphome/components/cst820/touchscreen/cst820.cpp [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+  // Update display dimensions if they were updated during display setup
+  this->x_raw_min_ = 0;
+  this->y_raw_min_ = 0;
+  if (this->swap_x_y_) {
+    this->x_raw_max_ = this->get_height_();
+    this->y_raw_max_ = this->get_width_();
+  } else {
+    this->x_raw_max_ = this->get_width_();
+    this->y_raw_max_ = this->get_height_();
+  };
```
Comment (nielsnl68) [inline esphome/components/cst820/touchscreen/cst820.cpp]:
```suggestion
    if (this->display_ != nullptr) {
      if (this->x_raw_max_ == this->x_raw_min_) {
        this->x_raw_max_ = this->display_->get_width_internal();
      }
      if (this->y_raw_max_ == this->y_raw_min_) {
        this->y_raw_max_ = this->display_->get_height_internal();
      }
    }
```

File: esphome/components/cst820/touchscreen/cst820.cpp [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+  if (err) {
+    ESP_LOGI(TAG, "Failed to read register 0x3");
+    return;
+  };
+
+  u_int16_t x, y;
+  x = (((data[0] & 0x0f) << 8) | data[1]);
+  y = (((data[2] & 0x0f) << 8) | data[3]);
+
+  set_raw_touch_position_(0, x, y);
```
Comment (nielsnl68) [inline esphome/components/cst820/touchscreen/cst820.cpp]:
```suggestion
  set_raw_touch_position_(fingerindex, x, y);
```

File: esphome/components/cst820/touchscreen/cst820.h [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+#define REG_CST820_IRQCONTROL       0xFA
+#define REG_CST820_AUTOSLEEP        0xFE
+
+using namespace touchscreen;
+
+class CST820Touchscreen : public Touchscreen, public i2c::I2CDevice {
+ public:
+  void setup() override;
+  void dump_config() override;
+  float get_setup_priority() const override;
```
Comment (nielsnl68) [inline esphome/components/cst820/touchscreen/cst820.h]: Im not sure of we should add the `get_setup_priority()` override. None of the existing touchscreen drivers (except the *XPT2046*) has it set atm.  So i decided to remove it also from the  *XPT2046* driver.

**Review by nielsnl68 — CHANGES_REQUESTED** [review]
Starting to look good. Still 3 small changes to make. 
Thanks for your effort.

File: esphome/components/cst820/touchscreen/cst820.cpp (hunk 16 above)
Comment (dirkjankrijnders) [reply esphome/components/cst820/touchscreen/cst820.cpp]: Of course, fixed

File: esphome/components/cst820/touchscreen/cst820.h (hunk 17 above)
Comment (dirkjankrijnders) [reply esphome/components/cst820/touchscreen/cst820.h]: I kept as my example code had it, will be removed here as well.

File: esphome/components/cst820/touchscreen/cst820.cpp (hunk 15 above)
Comment (dirkjankrijnders) [reply esphome/components/cst820/touchscreen/cst820.cpp]: The _internal are member of `DisplayBuffer`, but `this->display_` is of class `Display`. How to resolve that neatly?

**Comment by nielsnl68** [comment]
BTW, are you sure this is a single touch display?

**Comment by dirkjankrijnders** [comment]
> BTW, are you sure this is a single touch display?

No, I think it is multitouch, I get `fingerindex = 2` when I use two or more touches. I have never seen `fingerindex = 3`. However, the X en Y register only report 1 coordinate, the first one, in my testing. However, looking through other implementations there seem to be more registers then the datasheet defines. I focussed on getting single stable first, before diving into undocumented features. I do plan to revisit multi touch in the future though.

**Comment by pixelwave** [comment]
> Added support for the CST860 Touch screen controller as found on 2432S024C boards.

I have the ESP32-2432S024C. I think it has the CST820 touch controller? Maybe a typo because the headline says 820 as well? Or maybe I am missing something ...

Anyways I would like to get the touch functionality working. I got the screen displaying stuff with this code:

```
color:
  - id: my_red
    red: 100%
    green: 3%
    blue: 5%

spi:
 clk_pin: GPIO14
 mosi_pin: GPIO13
 miso_pin: GPIO12

display:
  - platform: ili9xxx
    model: TFT 2.4
    cs_pin: GPIO15
    dc_pin: GPIO2
    rotation: 90
    id: tft_ha
    lambda: |-
      it.fill(my_red);
``` 

How do I get the touch now working / what files do I have to include in my ESPHome build? Sorry I am still learning ESPHome ... quite new to the syntax. If somebody could give me support that would be awesome.

For example can I include the changed files as custom components like this or do I need to compile my own ESPHome binary:
```
esphome:
  includes:
    - cst820/binary_sensor.py
    - ...
```

**Comment by dirkjankrijnders** [comment]
Hi,

There is another PR that supports this chip is a more generic sense: https://github.com/esphome/esphome/pull/5941. That PR sees more active development. 
I plan to support that PR instead of this one, but I have no working hardware at the moment to test either one.

**Comment by pixelwave** [comment]
Thanks @dirkjankrijnders for the feedback. Will look into this PR. Do you know if the CST820 is compatible with the CST816?

**Comment by clydebarrow** [comment]
Closing in favour of #5941

**Comment by clydebarrow** [comment]
> Thanks @dirkjankrijnders for the feedback. Will look into this PR. Do you know if the CST820 is compatible with the CST816?

The CST820 is supported in ESPHome now - just merged into `dev` branch, as part of the `cst816` component. Beta goes out this week I believe.'

Any testing you can do will be appreciated.

**Comment by pixelwave** [comment]
Awesome - thanks @clydebarrow !

Would love to do some testing. But need help how to use the dev branch / integrate the touch controller...  I started this way: https://github.com/esphome/esphome/pull/5941#issuecomment-1963471884

**Comment by clydebarrow** [comment]
> Awesome - thanks @clydebarrow !
> 
> Would love to do some testing. But need help how to use the dev branch / integrate the touch controller... I started this way: [#5941 (comment)](https://github.com/esphome/esphome/pull/5941#issuecomment-1963471884)

You can install the ESPHome (dev) addon in HA. That will build against the dev branch. Or wait for the beta in a few days and use the beta addon.

**Pull request closed without merging**